Add AMI set entity and its create and delete endpoints

anarute opened the pull request:
I was not sure about its endpoints, if I am using the correct AMI Set properties.

jhford commented on src/ami-set.js:
```diff
@@ -0,0 +1,23 @@
+
+  properties: {
+
+    amiSetId: base.Entity.types.String,
```
since we're in the AmiSet class, let's call this just 'id'

jhford commented on src/ami-set.js:
```diff
@@ -0,0 +1,23 @@
+  properties: {
+
+    amiSetId: base.Entity.types.String,
+    amis: base.Entity.types.JSON,
```
Can you write a comment here that describes what this object should look like?  It's really helpful to know what to expect in the data when coding against this Entity in the future.

djmitche commented on src/api-v1.js:
```diff
@@ -580,6 +580,105 @@ api.declare({
+  deferAuth: true,
+  scopes: [['aws-provisioner:manage-ami-set:<amiSetId>']],
+  input: 'create-ami-set-request.json#',
+  output: 'get-ami-set-response.json#',
```
These (input and output) point to the schemas we talked about today.  You can just omit these lines for the moment, and then return to add schemas in a later pull request.

djmitche commented on src/api-v1.js:
```diff
@@ -580,6 +580,105 @@ api.declare({
+  let input = req.body;
+  let amiSet = req.params.amiSetId;
+
+  input.lastModified = new Date();
```
A `lastModified` property is a good idea, but you'll need to include it in `src/ami-set.js` as well.

djmitche commented on src/api-v1.js:
```diff
@@ -580,6 +580,105 @@ api.declare({
+  // Publish pulse message
+  await this.publisher.amiSetCreated({
+    amiSet: amiSetId,
+  });
```
I don't think we need to send a pulse message on AMI set creation.  It doesn't hurt, and we could add it back in later.

For WorkerTypes, this message is required so that the provisioner backend can look at the new workerType and start creating EC2 instance for it.  For new AmiSets, there's no need for the backend to do anything.

jhford replied:
We don't actually consume any pulse messages in the provisioner.  We scan the worker type table on each iteration, since we need to load all the worker types anyway to see what the max/min capacity, scaling ratio and instance types.  We could maintain that state in the backend process, but that's not there yet and would be a fairly substantial code change.  I don't know if the Azure overhead (tech or billing wise) is worth the trouble.

That said, I don't think that we really need to have pulse messages for AmiSets.  The important Pulse messages are for WorkerType updates, since those are (well, should be) used to gracefully shutdown old instances when we update the worker type definition.

djmitche commented on src/api-v1.js:
```diff
@@ -580,6 +580,105 @@ api.declare({
+  deferAuth: true,
+  scopes: [['aws-provisioner:manage-ami-set:<amiSetId>']],
+  input: undefined,  // No input
+  output: undefined,  // No output
```
You can just omit these

djmitche commented on src/api-v1.js:
```diff
@@ -580,6 +580,105 @@ api.declare({
+  let that = this;
+  let amiSet = req.params.amiSetId;
+
+  if (!req.satisfies({amiSet: amiSetId})) { return undefined; }
```
Why `undefined`?

anarute replied:
ops, sorry, I think I misunderstood this from worker-type.js

djmitche replied:
An understandable confusion -- `WorkerType.create` actually overrides the `Entity.create` method to give it a different function signature.  That's not necessary for AmiSet.

jhford commented on src/main.js:
```diff
@@ -11,6 +11,8 @@ let base = require('taskcluster-base');
 
 let workerType = require('./worker-type');
 let secret = require('./secret');
+let workerState = require('./worker-state');
```
You should rebase your patch to the latest version of the upstream master branch.  There was a change that I landed today that deletes the WorkerState entity, and it looks like you're adding it back here.

jhford commented on src/main.js:
```diff
@@ -59,6 +61,30 @@ let load = base.loader({
+    },
+  },
+
+  amiSet: {
```
We have a convention of anything that's a `taskcluster-base.Entity` using class casing, so please call this AmiSet here, and the other places that you are referring to the thing returned by `AmiSet.setup` instead of an instance of an `AmiSet`.

djmitche replied:
This should be capitalized (`AmiSet`).

djmitche commented on src/main.js:
```diff
@@ -59,6 +61,30 @@ let load = base.loader({
+    setup: async ({cfg}) => {
+      let amiSet = amiSet.setup({
+        account: cfg.azure.account,
+        table: cfg.app.workerStateTableName,
```
I think you'll need a different table name for ami sets :)

djmitche commented on src/ami-set.js:
```diff
@@ -0,0 +1,23 @@
+ * by its AWS region.
+ */
+
+let amiSet = base.Entity.configure({
```
This should be capitalized (`AmiSet`).  The idea is that this is a class, so just like in Python we use an initial capital letter.

djmitche commented on src/ami-set.js:
```diff
@@ -0,0 +1,23 @@
+    amis: base.Entity.types.JSON,
+
+  },
+});
```
The content of this module looks OK, but it doesn't export anything.  There's some information on exports here:

  https://www.sitepoint.com/understanding-module-exports-exports-node-js/

In this case, we want `require('./ami-set')` to return the configured entity, just like for `require('./worker-type')`.

anarute replied:
oh, right! so, I think I also need an AmiSet.create function, right?

djmitche replied:
You get that one for free -- `base.Entity.configure(..)` creates a class for you with a few methods (`create`, `update`, `remove`, etc.).  Sadly there's basically no documentation for that (https://bugzilla.mozilla.org/show_bug.cgi?id=1278385).

djmitche commented on src/main.js:
```diff
@@ -59,6 +61,30 @@ let load = base.loader({
+      });
+      return WorkerState;
+    },
+  },
```
I think this may be a merge error?  @jhford just removed the WorkerState entity class.  Awesome that you've rebased your changes against the latest master, by the way!

anarute replied:
yes, it is a merge error! probably it was from the commit last week and I didn't see this part was removed. Now I see the importance of sending small patches as soon as I can :)

jhford commented on src/api-v1.js:
```diff
@@ -580,6 +580,105 @@ api.declare({
+  ].join('\n'),
+}, async function (req, res) {
+  let input = req.body;
+  let amiSet = req.params.amiSetId;
```
I probably would call this `amiSetId` since you'll be creating an Entity and probably want this name for that, instead of `aSet`

djmitche commented on src/main.js:
```diff
@@ -124,13 +150,15 @@ let load = base.loader({
-    setup: async ({cfg, WorkerType, Secret, ec2, stateContainer, validator, publisher, influx}) => {
+    requires: ['cfg', 'WorkerType', 'WorkerState', 'amiSet', 'Secret', 'ec2', 'validator', 'publisher', 'influx'],
+    setup: async ({cfg, WorkerType, WorkerState, amiSet, Secret, ec2, validator, publisher, influx}) => {
+
```
As above, I don't think you want WorkerState in here (@jhford please correct me if I'm wrong!)

jhford replied:
Nope, she doesn't.  I think that a merge went a little wrong, because I just today deleted the WorkerState on the master branch.

djmitche commented on src/main.js:
```diff
@@ -124,13 +150,15 @@ let load = base.loader({
       let router = await v1.setup({
         context: {
           WorkerType: WorkerType,
+          amiSet: amiSet,
```
This too should be capitalized (`AmiSet`)

jhford commented:
@anarute really good first pass!  Most of the things that are needed are here.  It seems like you and @djmitche are talking about json-schema for the AmiSets, which is a great idea.  Otherwise, looking great!

djmitche commented:
OK that was a lot of comments, but this looks good!

If I remember correctly, you were able to run the tests for the provisioner.  Do they still run after your changes?

You will also need some tests in `test-src/manage_ami_set_test.js`, similar to `test-src/manageworkertype_test.js` (but again, much shorter, since so far AMI sets are a lot less complicated than worker types).  If it would help you out, I can write up a test for one of the API methods as an example.

anarute commented:
Thank you @djmitche and @jhford for the comments and feedback! It was really helpful to understand things better. I *think* (if I have not added more things to fix) that I've covered most part of the issues, could you please review it again?

jhford commented on src/api-v1.js:
```diff
@@ -584,41 +584,39 @@ api.declare({
 }, async function (req, res) {
   let input = req.body;
-  let amiSet = req.params.amiSetId;
+  let AmiSetId = req.params.id;
```
AmiSet is a 'class', but this isn't referring to the class.  This is referring to a string and so should be amiSetId or just id... either one is fine.

jhford commented on src/api-v1.js:
```diff
@@ -584,41 +584,39 @@ api.declare({
 
   // Create amiSet
-  let aSet;
+  let amiSetId;
```
this *will* refer to an AmiSet instance, so let's call it amiSet.

Basically, if a thing refers to a class (or in JS terms a constructor), it should be Caps

Here are some examples:
```javascript
class ExampleClass { }

function ExampleConstructor () { }

let ExampleClass = require('./lib/example-class').ExampleClass;
let exampleInstance = new ExampleClass();

let a = 'john';
```

jhford commented on src/main.js:
```diff
@@ -61,27 +60,15 @@ let load = base.loader({
+      let AmiSet = AmiSet.setup({
         account: cfg.azure.account,
-        table: cfg.app.workerStateTableName,
+        table: cfg.app.amiSetTableName,
```
We'll need to add this variable to the config file `config.yml`

jhford commented on src/ami-set.js:
```diff
@@ -8,16 +8,37 @@ const KEY_CONST = 'ami-set';
+    id: base.Entity.types.String,
+    /* This is a JSON object which contains the AMIs of an AMI set keyed by
+     * their virtualization type and region. It is in the shape:
+     * {
```
This is not specific to your patch, but I'm sort of wondering if a structure like
```json
{
   "us-west-1": {
      "hvm": "ami-111",
      "pv": "ami-222"
  },
   "us-east-1": {
      "hvm": "ami-111",
      "pv": "ami-222"
  }
}
```
or
```json
[
   {
      "region": "us-west-1",
      "hvm": "ami-111",
      "pv": "ami-222"
   },
   {
      "region": "us-east-1",
      "hvm": "ami-111",
      "pv": "ami-222"
   }
]
```
The reason for flipping the hvm/pv around is that the structure goes from most general (region) to more specific (virtualization style) to most specific (ami id).

Whether we have a list of objects or a straigh mapping is something that you'll have to pick.  A list of objects is really nice because you can do really neat things with `Array.prototype.map` and `Array.prototype.filter`.  A mapping is really nice because it's easier to address the values and it is impossible to have duplicates.  Error checking with the object will be easier, but the list of objects is more in line with what we do elsewhere in the provisioner and probably what I'd prefer to see.

djmitche replied:
I tend to agree regarding the list of objects (the second option).  That will also mean we don't need to change our schema when we add a new region.  I also agree that we don't need to solve it in this patch.

jhford commented on src/main.js:
```diff
@@ -124,13 +137,15 @@ let load = base.loader({
   api: {
-    requires: ['cfg', 'WorkerType', 'Secret', 'ec2', 'stateContainer', 'validator', 'publisher', 'influx'],
-    setup: async ({cfg, WorkerType, Secret, ec2, stateContainer, validator, publisher, influx}) => {
+    requires: ['cfg', 'WorkerType', 'AmiSet', 'Secret', 'ec2', 'validator', 'publisher', 'influx'],
```
There's another merge issue here.  stateContainer needs to be in both the requires list and the unpacked variable assignments in the setup function signature, but it's not.

jhford commented on src/api-v1.js:
```diff
@@ -580,6 +580,94 @@ api.declare({
 
 api.declare({
   method: 'put',
+  route: '/ami-set/:amiSetId',
```
let's go with just `:id` here as well.  When inside the managing functions for a thing itself, i find it easier to just refer to it as 'id' or something, and use more descriptive names for the ids of things that are referenced as a part of a thing.

jhford commented on src/api-v1.js:
```diff
@@ -580,6 +580,94 @@ api.declare({
+  input.lastModified = new Date();
+
+  // Authenticate request with parameterized scope
+  if (!req.satisfies({amiSet: AmiSetId})) {
```
the property that is checked against (here `amiSet`) needs to match the parameterized name in the scopes list (here `amiSetId`, but should be `id`).

After renaming the `amiSetId` variables and scopes to `id`, the object to use here would be `{id: id}`.  In es6+, if your property name and value are derived from the save variable, e.g. `let x = 1; {x: x}` then you can just do `let x = 1; {x}.

jhford commented on src/api-v1.js:
```diff
@@ -580,6 +580,94 @@ api.declare({
+  let amiSet = req.params.amiSetId;
+
+  try {
+    await this.amiSet.remove({amiSet: amiSetId}, true);
```
Remember that you're passing in the AmiSet entity class as a context variable to this function in `src/main.js`.  This name here needs to be the name of the context object key, since what we do is take the 'context' object and assign all those values as properties on the `this` in this scope.  In other words, you're looking for `this.AmiSet.remove`.

jhford commented on src/main.js:
```diff
@@ -124,13 +137,15 @@ let load = base.loader({
       let router = await v1.setup({
         context: {
           WorkerType: WorkerType,
+          AmiSet: AmiSet,
```
Regarding my comment about `this.AmiSet.remove`, this is the place where you're creating the value `this.AmiSet`.  This is magic done by the library, but fyi.

jhford commented:
I'm not sure what the scope of this PR is, but I noticed that there isn't a view or update endpoint for the `AmiSets`.  If that's outside the scope of the PR as you've discussed with @djmitche , it's fine to hold off on that for the time being, but at some point we will need to add these :)

jhford commented on src/main.js:
```diff
@@ -59,6 +60,18 @@ let load = base.loader({
+  AmiSet: {
+    requires: ['cfg'],
+    setup: async ({cfg}) => {
+      let AmiSet = AmiSet.setup({
```
slight bug here.  You are declaring a new variable, `AmiSet`, but assigning the first value to something that uses the variable declared here.  This alone will cause
```
localhost:~/taskcluster/aws-provisioner $ node
> let x = x;
ReferenceError: x is not defined
    at repl:1:31
    at REPLServer.defaultEval (repl.js:252:27)
    at bound (domain.js:287:14)
    at REPLServer.runBound [as eval] (domain.js:300:12)
    at REPLServer.<anonymous> (repl.js:417:12)
    at emitOne (events.js:82:20)
    at REPLServer.emit (events.js:169:7)
    at REPLServer.Interface._onLine (readline.js:210:10)
    at REPLServer.Interface._line (readline.js:549:8)
    at REPLServer.Interface._ttyWrite (readline.js:826:14)
```
What you're looking for here is
```javascript
      let AmiSet = amiSet.setup({ 
```
because you imported the `AmiSet` 'class' using the name `amiSet` above, where you `require()` the file.

jhford commented on test-src/manageamiset_test.js:
```diff
@@ -0,0 +1,58 @@
+
+  var id = slugid.nice();
+  var amiSetDefinition = makeAmiSet();
+  var amiSetChanged = _.clone(amiSetDefinition);
```
Without an update endpoint, this variable is not really useful.  For the worker type tests, we have this changed variable where we basically create a second version of the tested definition, with one value changed.  We then use the update method to verify that this expected value matches the one out of the api/table storage.  For now, we should either delete this variable or we should have the update logic implemented.

djmitche commented:
Yes, we should have "get" and "list" endpoints too.  I feel silly for not noticing that! :)

djmitche commented:
From talking to @jhford and @anarute today, summarizing the remaining work:
 * get and list endpoints, plus tests
 * schema validation for the `[{region, hvm, pv}, ..]` format

jhford commented:
Oh, and also an update endpoint.

djmitche commented on src/api-v1.js:
```diff
@@ -632,6 +632,67 @@ api.declare({
+    } else {
+      throw err;
+    }
+  }
```
This never returns `amiSet`..

djmitche commented on src/api-v1.js:
```diff
@@ -632,6 +632,67 @@ api.declare({
+      debug(err.stack);
+    }
+    throw err;
+  }
```
It looks like elements of `list` here would have a different format than that returned from `getAmiSet`, since `listAmiSets` is using `item.ami.json()`.  The least surprising thing for a user of an API is if the two data structures have the same format.  Using JSON schemas will help enforce this.

jhford replied:
I would almost say that we should just return a list of names, since that's what we do for the worker types.

djmitche replied:
It's done either way in various places in TaskCluster.  I think the key factors are how much work it is to fetch all of the data, and how long the response might get.  In this case, scanning the table to get the names is no less work than scanning the table to get all of the AMI set content; and the resulting response is likely to be a fairly small fixed length (dozens of ami sets), so it's best to return the whole AMI set.

anarute replied:
Should I return `amiSet.amis.json()` in the getAmiSet then?

djmitche replied:
That would refer to a `json()` method on the `amis` property of the azure entity.  Looking in `ami-set.js`, no such method is defined.

It probably makes sense to define an `AmiSet.json()` method which will return the JSON-able value for the whole AMI set -- including `lastModified` and `id`.  Then you can call that method in both getAmiSet and listAmiSets.

djmitche replied:
This method is looking great now.

jhford replied:
The provisioner mostly uses list of strings for list methods.  I'd like to stick to that here unless there's a compelling reason not to.  That would also make having a json-schema for this endpoint pretty quick to write up.

jhford commented on src/ami-set.js:
```diff
@@ -41,4 +41,29 @@ let AmiSet = base.Entity.configure({
+ */
+AmiSet.listAmiSets = async function () {
+
+  let amislist = [];
```
Normally, we'd camel case this into `amisList`, but given that a list is an aggregation(in this case, a list), I'd normally use a singular for the first part, so `amiList`, but since we're talking about ami sets and not amis, I'd call it `amiSetList`.

djmitche commented on schemas/create-ami-set-request.yml:
```diff
@@ -0,0 +1,39 @@
+      pv:
+        type: string
+        description: |
+          The AMI that uses PV virtualization type
```
you'll want

```
additionalProperties: false
required:
  - region
  - hvm
  - pv
```
here, too.

djmitche commented on schemas/get-ami-set-response.yml:
```diff
@@ -0,0 +1,39 @@
+      pv:
+        type: string
+        description: |
+          The AMI that uses PV virtualization type
```
Same (`additionalProperties`/`required`) here, too.

jhford replied:
wait... why?  pv is a type of string, it doesn't have properties... am I missing something?

djmitche replied:
Oh, indentation is a bit funny here, too -- it says `amis` has properties, rather than the items in `amis`.  My comment was indicating that the region, hvm, and pv properties should be limited and all required.

jhford replied:
ahh right, so just intending it a bit.  `amis` is an array, so it shouldn't have any properties.  I'm surprised this schema works

jhford replied:
http://jsonschemalint.com/draft4/ is your friend.  I patched it to support YAML serialised schemas and documents

djmitche commented on src/api-v1.js:
```diff
@@ -580,6 +580,213 @@ api.declare({
+  input.lastModified = new Date();
+
+  // Authenticate request with parameterized scope
+  if (!req.satisfies({id: id})) {
```
(nit) `{id}`

anarute replied:
I didn't understand, is `{id}` a short form for `{id: id}`?

djmitche replied:
Yep: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer
```
With ECMAScript 2015, there is a shorter notation available to achieve the same:

var a = "foo", 
    b = 42, 
    c = {};

// Shorthand property names (ES6)
var o = { a, b, c };
```

djmitche commented on src/api-v1.js:
```diff
@@ -580,6 +580,213 @@ api.declare({
+    for (let key of Object.keys(input)) {
+      amiS[key] = input[key];
+      amiS.lastModified = modDate;
+    }
```
There's really only one key of interest here (`amis`).  The `lastModified` doesn't need to be copied from input (and `input.lastModified` doesn't need to be set, above, either).  So while this works, it's a bit over-complicated.

anarute replied:
should I just remove this logic for now?

djmitche replied:
Simplify, yes (basically just `amiS.amis = input.amis`)

djmitche commented on src/api-v1.js:
```diff
@@ -580,6 +580,213 @@ api.declare({
+  output: 'get-ami-set-response.json#',
+  deferAuth: true,
+  scopes: [
+    ['aws-provisioner:view-ami-set:<amiSetId>'],
```
Here, `getAmiSet` requires `aws-provisioner:view-ami-set:<id>`, but `listAmiSets` requires no scopes.  It's tricky to enforce scopes in a list method, and I don't see a great reason to hide AMI sets from public view, so I think we should just allow anyone to call `getAmiSet`, too (meaning, remove the `scopes:` here and the `req.satisifies` below).

djmitche commented on test-src/manageamiset_test.js:
```diff
@@ -0,0 +1,77 @@
+
+  // var id = slugid.nice();
+  var amiSetDefinition = makeAmiSet();
+  var amiSetChanged = _.clone(amiSetDefinition);
```
It doesn't look like this AMI set is actually different from the original.  Maybe add a new AMI to it, just to be more realistic?

djmitche commented on test-src/manageamiset_test.js:
```diff
@@ -0,0 +1,77 @@
+    } catch (err) {
+      assume(err.statusCode).equals(404);
+    }
+  });
```
I don't see a test for `listAmiSets`..

djmitche commented:
John's review is the one that counts, but as far as I can see this is looking good with just a few more minor things to fix:

 * tests for `listAmiSets`
 * actually change the AMI set in tests
 * schema fixes
 * remove scopes for `getAmiSets`

Once you get going writing tools and backend support for this, you'll probably find some other things to change, but that's easy to do with another PR.

anarute commented:
Cool! I'm working on this changes and I'll rebase these commits to squash the "fix" ones when I finish them

djmitche commented on test-src/manageamiset_test.js:
```diff
@@ -61,6 +67,10 @@ describe('AMI Set api', () => {
+  it('should return a list of AMI sets', async () => {
+    await client.listAmiSets();
+  });
+
```
does it actually return a list of AMIs? :)

djmitche replied:
Per out discussion, this should use `assume(amiSets).to.deeply.equal(..)`.

jhford commented on src/ami-set.js:
```diff
@@ -0,0 +1,73 @@
+  try {
+    await base.Entity.scan.call(this, {}, {
+      handler: function(item) {
+        amiSetList.push(item.json());
```
If this is a listing method, we should be pushing the item.id property instead of the complete item in a json-serializable format.  If the goal is a load-all method, let's call it `loadAll()` to match the worker type [here](https://github.com/taskcluster/aws-provisioner/blob/master/src/worker-type.js#L249)

djmitche replied:
I believe we already had some discussion of this in the PR?  Many of the list methods in TC services return the entire object, e.g., [auth.listClients](https://docs.taskcluster.net/reference/platform/auth/api-docs#listClients).  The reason *not* to do so is if it requires additional queries or a lot more bandwidth.  Neither is the case here.  We have the additional information about each AMI set returned from Azure, so why not provide it to the user?

jhford commented on src/api-v1.js:
```diff
@@ -580,6 +580,197 @@ api.declare({
+      'id',
+      'amis',
+    ].every((key) => {
+      return _.isEqual(amiSet[key], input[key]);
```
I can't remember off hand, but we should ensure that `_.isEqual` is a deep equality operator, since an amiSet is an object which collects other objects.

djmitche replied:
[It is.](https://lodash.com/docs#isEqual)

jhford commented on src/api-v1.js:
```diff
@@ -580,6 +580,197 @@ api.declare({
+api.declare({
+  method: 'get',
+  route: '/ami-set/:id',
+  name: 'getAmiSet',
```
The `WorkerType` entity uses `name:  'workerType'`.  Can we follow that pattern here?

jhford commented on src/api-v1.js:
```diff
@@ -580,6 +580,197 @@ api.declare({
+
+  let amiSet = await this.AmiSet.load({id: id});
+
+  await amiSet.modify(function(amiS) {
```
the name `amiS` sort of implies that it's a plural of amis but with confusing captalization.  Instead, since we're passing in an instance of an AmiSet, let's use the `amiSet` variable name.

jhford commented on test-src/manageamiset_test.js:
```diff
@@ -0,0 +1,144 @@
+  });
+
+  beforeEach(async () => {
+    await main('tableCleaner', {process: 'tableCleaner', profile: 'test'});
```
We should add the `AmiSet` entity to the `tableCleaner` component in `src/main.js` so that this table is actually cleaned out.

jhford commented on test-src/manageamiset_test.js:
```diff
@@ -0,0 +1,144 @@
+    amiSet = await client.getAmiSet(id);
+    lastModified = amiSet.lastModified;
+    assume(await amiSet).to.deeply.equal({
+      amis: [
```
we use mock objects in a single file elsewhere to make it easier to update the format of the test objects without having to completely rewrite the test.  In this case, the objects are simple enough that I don't know if it's worth doing here, but please take a look at the `test-src/mock*.js` files just to see how they're structured.  Basically, we have a default that is returned, but you can overwrite the values you want overwritten.  Not critical here, but worth looking at.

anarute commented:
@jhford I fixed the things you've mentioned:
- added AmiSet table to tablecleaner;
- fixed naming patterns;
- cleaned the amiSet tests using only data from mockamiset_test.js - except data for comparison, like here https://github.com/taskcluster/aws-provisioner/pull/85/files#diff-bf33c71f7b5151be915935e33c040095R66
- changed `listAmiSets` method to return the names of all amiSets, not the objects. Should I create a `loadAll()` method for amiSets?

djmitche commented:
If we need `loadAll()`, it's easy enough to add in a new PR :)

jhford commented:
So there's a problem here.  The create ami set endpoint has a json-schema that forces there to be an ID parameter in the request body.  The problem is that we already have that specified in the parameters of the api method.  The create schema should only have the amis object.

jhford commented:
Ok.  So I think the only issue left here is removing the createAmiSet json-schema allowing an ID value.  The only place we should be supplying the id value is the parameter that we bake into the `/ami-set/:id` route.

I think with that change, i'm 99% sure that we're ready to merge!

anarute commented:
@jhford I removed the id property from `createAmiSet` schema, could you review the code again, please?

jhford commented on schemas/create-ami-set-request.yml:
```diff
@@ -0,0 +1,39 @@
+        - region
+        - hvm
+        - pv
+  lastModified:
```
Sorry that I didn't see this before, but we also should not have last modified as an allowed field in the request.  we will generate this value inside the api handler, so forcing a value here that we won't end up using isn't ideal.

jhford commented on src/api-v1.js:
```diff
@@ -580,6 +580,197 @@ api.declare({
+  let input = req.body;
+  let id = req.params.id;
+
+  input.lastModified = new Date();
```
We shouldn't really be setting properties on the request body object.  Also, this value isn't actually used in the `AmiSet.create` method call below

jhford commented on src/api-v1.js:
```diff
@@ -580,6 +580,197 @@ api.declare({
+    amiSet = await this.AmiSet.create({
+      id: id,
+      amis: input.amis,
+      lastModified: new Date(),
```
here's where we're actually using this value

anarute replied:
you're right, I passed that out among all the changes! thanks for noticing

jhford commented on test-src/manageamiset_test.js:
```diff
@@ -0,0 +1,107 @@
+var slugid = require('slugid');
```
random nit, this file uses `let` and `var`.  We should probably chose one and stick to it.

The difference in case you're unfamiliar is that var is in scope through the entire function.  If you use var in an `if`, the variable is in scope through the entire function.  Let is block scope, if you use let in an `if`, it's only in scope inside that `if` and those blocks (e.g. code inside of `{ }`) that are nested.

Let is better than var!  Also, if you're really curious do a search about 'variable hoisting' for more info on how var works.

anarute replied:
thanks for the explanation!

jhford commented on test-src/manageamiset_test.js:
```diff
@@ -0,0 +1,107 @@
+    debug('### Load amiSet (again)');
+    amiSet = await client.amiSet(id);
+    lastModified = amiSet.lastModified;
+    assume(await amiSet).to.deeply.equal({
```
we don't need to await `amiSet` because we already await'd it above.

jhford replied:
Also, i think this is way outside the scope of this project, but one thing that I like to do is compare the amiSetChanged object itself.  Doing that does require that you handle the lastModified stuff, as you do correctly below.  Another option is to use Sinon to mock out the system clock.  That's really out of scope and what you have here is fine.  I might consider using `_.default` to grab the amis object from amiSetChanged
